### ISSUE TITLE
Avoid `SELECT MAX(control)` queries

### ISSUE THREAD

**Issue by nicolagi** [pgpartman/pg_partman#660]
I see that pg_partman runs `SELECT MAX(control)` to find what is the "next" partition to create.
That's expensive when a partition is large and there is no index on the control column (a situation I find myself in).
I wonder why not just trust the partition boundaries (perhaps only if `infinite_time_partitions` is true).
E.g., for a monthly-partitioned table `foo`, if `foo_p20240501` is present for values between first and last of May inclusive, assume new (monthly) partitions will be starting from `foo_p20240601` onwards.
I do see the need to still scan the DEFAULT table, in case there's recent data there that would conflict with new partitions.

IIRC, the current strategy also makes transitioning from monthly to, say, weekly partitions, hard. E.g., the `SELECT MAX(control)` method entails trying to create `foo_p20240508`, etc. which would conflict with the existing partition.

**Comment by keithf4** [comment]
That's a good suggestion to revisit that method now that trigger-based partitioning is no longer supported. I'll definitely have a closer look during the next version's code review. Thank you!

Also curious on not having the index there. Is the control column purely for retention purposes and the queries never use it as a condition?

**Comment by nicolagi** [comment]
> That's a good suggestion to revisit that method now that trigger-based partitioning is no longer supported. I'll definitely have a closer look during the next version's code review. Thank you!

Hopefully I'll be able to collaborate.

> Also curious on not having the index there. Is the control column purely for retention purposes and the queries never use it as a condition?

Precisely. (I was surprised as well at first.)

Forgive the possible off-topic, but it seems to me that if `run_maintenance` takes long, say, 5 minutes, then the initial DROP (when applying the retention policy) will acquire an exclusive lock on the table and block all queries to the table for the whole 5 minutes. If that reasoning is correct, it's all the more reason to avoid scanning if possible (and for me to index all control columns).

**Comment by keithf4** [comment]
Just reviewing when the drop happens, I saw that it's one of the first steps in maintenance, so I'm going to review why that couldn't possibly be later instead so any locks the drop takes aren't done until closer to the end. I cannot recall if that lock is exclusive for the entire partition set or not. It's definitely exclusive for that child table tho.

This is also a use-case for using `run_maintenance_proc()` vs standard `run_maintenance()`. The procedure commits after each partition set's maintenance so one's transactional holds do not interfere with any others.

**Comment by keithf4** [comment]
Further review, I don't think I can just use the child boundaries. partman also determines whether to actually make new tables or not based on whether data exists in the child tables. This is necessary for integer-based partitioning otherwise every maintenance run would always make new children. And I do the same for time just to be consistent with the way integer works, but it can be overridden with the `infinite_time_partitions` flag. 

But I think I can just do a `SELECT ... LIMIT 1` on the child table instead of the max. And if the flag to always make time partitions is set, it could possibly be optimized to just use the boundaries (using the `show_partition_info()` function). When I get around to working on it, I'll see. Or if you want to take a crack at it, I'm open to PRs.

**Comment by nicolagi** [comment]
> Just reviewing when the drop happens, I saw that it's one of the first steps in maintenance, so I'm going to review why 
> that couldn't possibly be later instead so any locks the drop takes aren't done until closer to the end. I cannot 
> recall if that lock is exclusive for the entire partition set or not. It's definitely exclusive for that child table 
> tho.

I did this test, in one session, I did

	BEGIN;
	DROP foo_p20240301;

and in another I did

	SELECT * FROM foo;

and the latter was blocked.
Then I issued a `ROLLBACK;` in the first session and the latter query returned.
Same with `INSERT INTO foo`.
OTOH, `SELECT * FROM foo_p20240501`, say, would not block, but that's of limited use.

> This is also a use-case for using `run_maintenance_proc()` vs standard `run_maintenance()`. The procedure commits after 
> each partition set's maintenance so one's transactional holds do not interfere with any others.

We don't use `run_maintenance_proc()`, but we do invoke `run_maintenance()` separately for each table, each in its own transaction.

> Another note on this. I may be able to remove the `SELECT max()` on the normal child tables, but if the option to 
> include data in the default table is enabled, there's no way to remove that `SELECT max()` since the default has no 
> edge boundaries. That default table inclusion option is not enabled by default, but just wanted to mention it.

Agreed, was trying to say in OP as well that we still must get the max from the default table.

> Further review, I don't think I can just use the child boundaries. partman also determines whether to actually make new 
> tables or not based on whether data exists in the child tables. This is necessary for integer-based partitioning 
> otherwise every maintenance run would always make new children. And I do the same for time just to be consistent with 
> the way integer works, but it can be overridden with the `infinite_time_partitions` flag. 
>
> But I think I can just do a `SELECT ... LIMIT 1` on the child table instead of the max. And if the flag to always make 
> time partitions is set, it could possibly be optimized to just use the boundaries (using the `show_partition_info()` 
> function). When I get around to working on it, I'll see. Or if you want to take a crack at it, I'm open to PRs.

Agree.
In OP I also was mentioning that perhaps the partition boundaries could be used only in the case `infinite_time_partitions` is true.
But your idea to use `SELECT ... LIMIT 1` seems best as that should be fast in all cases and you wouldn't have to special-case for `infinite_time_partitions`.
I'll try raise a PR.

**Comment by nicolagi** [comment]
> In OP I also was mentioning that perhaps the partition boundaries could be used only in the case `infinite_time_partitions` is true.

Actually, even with `infinite_time_partitions`, one has to find the latest non-empty partition if any (and default to current timestamp if all are empty). Otherwise, new partitions would be created at each run, as you have observed already for integer-based partitioning. So `LIMIT 1` seems to be the way to go.

**Comment by keithf4** [comment]
Not necessarily. `infinite_time_partitions` is based on what time it is `now()` when maintenance runs. It sees if there are premade tables based on `now()` and only makes them if there aren't any. For integer-based partitioning, there's no independent idea of what the "current" number is like there is with time and `now()`.

**Comment by nicolagi** [comment]
> Not necessarily. `infinite_time_partitions` is based on what time it is `now()` when maintenance runs.

AFAICS in [the current code](https://github.com/pgpartman/pg_partman/blob/master/sql/functions/run_maintenance.sql#L228-L243), only if `v_max_timestamp` is less than the current time or is null (i.e., the partition set is empty).

**Comment by keithf4** [comment]
Right. I guess my thought is that there wouldn't be any need to compare it to an actual value in the partition set anymore if we're just going to try and use the partition boundaries instead (unless the default table is included)

**Comment by reeyah** [comment]
I have a question here, since we look at the data and the premake value to create partitions instead can't we keep the control and now() and the last partition, because if there is no data in the table for a prolonged period, it won't create any new partitions but then it will also fail to insert any new data as it cannot find any relevant partition (I am excluding the default case from this scenario since I do not want the table as per my use case)

**Comment by keithf4** [comment]
If you're talking about the scenario where no new child tables are created for a long time and then someone goes and inserts new data expecting that new child table to be there, then that's what the `infinite_time_partitions` option is designed to handle. ID/int based partitioning cannot continuously make new child tables without new data, so I made the time-based partitioning work the same by default. Also, if no new data is being inserted into a table, I simply wanted to avoid needlessly continuing to make new tables for a partition set that may have been abandoned and cluttering up the system catalogs. But if people do want that behavior, I want them to set it intentionally.

**Comment by reeyah** [comment]
So the issue I faced is that I use `CALL partman.run_maintenance_proc();` periodically on my db. The detaching of outdated partitions happen as expected but no new partitions are created as per the premake value. I used range partitioning on the table with 7 days partition policy and 7 days retention policy. There is no new data entry in any of the tables for the last 15 days or so but when I tried inserting new data it throws error that it cannot find a suitable partition to enter the data. The infinite_time_partitions is already set to true yet the issue is same. Also, the default partition was already dropped since that is not part of my use case. So what is the solution here?

**Comment by keithf4** [comment]
Can you output the contents of `part_config` for this table and the `\d+` schema for it. This is the exact scenario that `infinite_time_partitions` is supposed to handle. It should keep making partitions forever even when no new data is inserted.

**Comment by reeyah** [comment]
```[
	{
		"parent_table": "public.vkyc",
		"control": "created_at",
		"partition_interval": "7 days",
		"partition_type": "range",
		"premake": 1,
		"automatic_maintenance": "on",
		"template_table": "partman.template_public_vkyc",
		"retention": "7 days",
		"retention_schema": null,
		"retention_keep_index": true,
		"retention_keep_table": true,
		"epoch": "none",
		"constraint_cols": null,
		"optimize_constraint": 30,
		"infinite_time_partitions": true,
		"datetime_string": "YYYYMMDD",
		"jobmon": true,
		"sub_partition_set_full": false,
		"undo_in_progress": false,
		"inherit_privileges": false,
		"constraint_valid": true,
		"ignore_default_data": true,
		"default_table": true,
		"date_trunc_interval": null
	}
]```

and table definition:
```CREATE TABLE vkyc (
    request_id character varying,
    user_id character varying,
    vendor character varying,
    state character varying,
    meta character varying,
    tenant character varying,
    created_at timestamp without time zone,
    updated_at timestamp without time zone,
    version bigint,
    flow character varying,
    CONSTRAINT vkyc_pkey PRIMARY KEY (request_id, created_at)
);

-- Indices -------------------------------------------------------

CREATE UNIQUE INDEX vkyc_pkey ON vkyc(request_id text_ops,created_at timestamp_ops);
CREATE INDEX vkyc_request_id_idx ON vkyc(request_id text_ops);
CREATE INDEX vkyc_user_id_idx ON vkyc(user_id text_ops);```

**Comment by reeyah** [comment]
@keithf4 can you help identify what is the issue here? From my observations and findings, the partition creation only works if there is any data in the existing partitions (also note that I have dropped the default partition after using create_parent() initially itself). However, the detaching is working perfectly in this case.

**Comment by keithf4** [comment]
Apologies for the lack of response on this issue. Been very busy with some other projects lately. I will be returning to partman work in the near future again and this will be the first issue I look into more once I do.

**Comment by keithf4** [comment]
> @keithf4 can you help identify what is the issue here? From my observations and findings, the partition creation only works if there is any data in the existing partitions (also note that I have dropped the default partition after using create_parent() initially itself). However, the detaching is working perfectly in this case.

So I made a working example from the code you gave. To simulate needing a future child table, I increased the premake value by 1 and maintenance seems to be working properly. Are you sure there are no errors being thrown when maintenance runs?
```
git660=# \d+ vkyc
                                                 Partitioned table "public.vkyc"
   Column   |            Type             | Collation | Nullable | Default | Storage  | Compression | Stats target | Description 
------------+-----------------------------+-----------+----------+---------+----------+-------------+--------------+-------------
 request_id | character varying           |           | not null |         | extended |             |              | 
 user_id    | character varying           |           |          |         | extended |             |              | 
 vendor     | character varying           |           |          |         | extended |             |              | 
 state      | character varying           |           |          |         | extended |             |              | 
 meta       | character varying           |           |          |         | extended |             |              | 
 tenant     | character varying           |           |          |         | extended |             |              | 
 created_at | timestamp without time zone |           | not null |         | plain    |             |              | 
 updated_at | timestamp without time zone |           |          |         | plain    |             |              | 
 version    | bigint                      |           |          |         | plain    |             |              | 
 flow       | character varying           |           |          |         | extended |             |              | 
Partition key: RANGE (created_at)
Indexes:
    "vkyc_pkey" PRIMARY KEY, btree (request_id, created_at)
    "vkyc_request_id_idx" btree (request_id)
    "vkyc_user_id_idx" btree (user_id)
Partitions: vkyc_p20240913 FOR VALUES FROM ('2024-09-13 00:00:00') TO ('2024-09-20 00:00:00'),
            vkyc_p20240920 FOR VALUES FROM ('2024-09-20 00:00:00') TO ('2024-09-27 00:00:00'),
            vkyc_p20240927 FOR VALUES FROM ('2024-09-27 00:00:00') TO ('2024-10-04 00:00:00'),
            vkyc_p20241004 FOR VALUES FROM ('2024-10-04 00:00:00') TO ('2024-10-11 00:00:00'),
            vkyc_p20241011 FOR VALUES FROM ('2024-10-11 00:00:00') TO ('2024-10-18 00:00:00'),
            vkyc_p20241018 FOR VALUES FROM ('2024-10-18 00:00:00') TO ('2024-10-25 00:00:00'),
            vkyc_p20241025 FOR VALUES FROM ('2024-10-25 00:00:00') TO ('2024-11-01 00:00:00'),
            vkyc_p20241101 FOR VALUES FROM ('2024-11-01 00:00:00') TO ('2024-11-08 00:00:00'),
            vkyc_p20241108 FOR VALUES FROM ('2024-11-08 00:00:00') TO ('2024-11-15 00:00:00'),
            vkyc_default DEFAULT

git660=# update partman.part_config set infinite_time_partitions = true;
UPDATE 1
git660=# update partman.part_config set premake = 5;
UPDATE 1
git660=# select partman.run_maintenance();
 run_maintenance 
-----------------
 
(1 row)

git660=# \d+ vkyc
                                                 Partitioned table "public.vkyc"
   Column   |            Type             | Collation | Nullable | Default | Storage  | Compression | Stats target | Description 
------------+-----------------------------+-----------+----------+---------+----------+-------------+--------------+-------------
 request_id | character varying           |           | not null |         | extended |             |              | 
 user_id    | character varying           |           |          |         | extended |             |              | 
 vendor     | character varying           |           |          |         | extended |             |              | 
 state      | character varying           |           |          |         | extended |             |              | 
 meta       | character varying           |           |          |         | extended |             |              | 
 tenant     | character varying           |           |          |         | extended |             |              | 
 created_at | timestamp without time zone |           | not null |         | plain    |             |              | 
 updated_at | timestamp without time zone |           |          |         | plain    |             |              | 
 version    | bigint                      |           |          |         | plain    |             |              | 
 flow       | character varying           |           |          |         | extended |             |              | 
Partition key: RANGE (created_at)
Indexes:
    "vkyc_pkey" PRIMARY KEY, btree (request_id, created_at)
    "vkyc_request_id_idx" btree (request_id)
    "vkyc_user_id_idx" btree (user_id)
Partitions: vkyc_p20240913 FOR VALUES FROM ('2024-09-13 00:00:00') TO ('2024-09-20 00:00:00'),
            vkyc_p20240920 FOR VALUES FROM ('2024-09-20 00:00:00') TO ('2024-09-27 00:00:00'),
            vkyc_p20240927 FOR VALUES FROM ('2024-09-27 00:00:00') TO ('2024-10-04 00:00:00'),
            vkyc_p20241004 FOR VALUES FROM ('2024-10-04 00:00:00') TO ('2024-10-11 00:00:00'),
            vkyc_p20241011 FOR VALUES FROM ('2024-10-11 00:00:00') TO ('2024-10-18 00:00:00'),
            vkyc_p20241018 FOR VALUES FROM ('2024-10-18 00:00:00') TO ('2024-10-25 00:00:00'),
            vkyc_p20241025 FOR VALUES FROM ('2024-10-25 00:00:00') TO ('2024-11-01 00:00:00'),
            vkyc_p20241101 FOR VALUES FROM ('2024-11-01 00:00:00') TO ('2024-11-08 00:00:00'),
            vkyc_p20241108 FOR VALUES FROM ('2024-11-08 00:00:00') TO ('2024-11-15 00:00:00'),
            vkyc_p20241115 FOR VALUES FROM ('2024-11-15 00:00:00') TO ('2024-11-22 00:00:00'),
            vkyc_default DEFAULT
```

**Comment by keithf4** [comment]
The original topic this issue was opened for has been addressed in 5.2.0. Thanks again for the suggestion on performance improvements.

@reeyah If you're still having issues, please open up a new, separate issue to continue discussing there. Thank you!
